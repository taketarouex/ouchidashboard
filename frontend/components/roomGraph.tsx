import React from 'react';
import CircularProgress from '@material-ui/core/CircularProgress'
import { LineChart } from 'recharts'
import useSWR from 'swr'
import dayjs from 'dayjs'

type Log = {
  value: number
  updatedAt: Date
}

const fetchLogs: (url: string) => Promise<Log[]> = (url) => fetch(url).then(
  res => {
    if (!res.ok) {
      const error = new Error('An error occurred while fetching the data.')
      throw error
    }
    return res.json()
  }
)

const useRoomLog = ({ roomName, logType, start, end }: { roomName: string, logType: string, start: dayjs.Dayjs, end: dayjs.Dayjs }) => {
  const startISO = start.toISOString()
  const endISO = end.toISOString()
  const { data, error } = useSWR(`/api/rooms/${roomName}/logs/${logType}?start=${startISO}&end=${endISO}`, fetchLogs)

  return {
    data: data,
    isLoading: !error && !data,
    isError: error
  }
}

export const RoomGraph = ({ roomName, logType, start, end }: { roomName: string, logType: string, start: dayjs.Dayjs, end: dayjs.Dayjs }) => {
  const { data, isLoading, isError } = useRoomLog({ roomName, logType, start, end })
  if (isLoading) return <CircularProgress />
  if (isError) return <div>error</div>
  return (
    <LineChart
      width={400}
      height={400}
      data={data}
    >
    </LineChart>)
}