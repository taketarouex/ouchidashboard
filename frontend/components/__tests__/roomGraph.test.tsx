import React from 'react';
import { render, waitForElementToBeRemoved, screen } from '@testing-library/react'
import { RoomGraph } from '../roomGraph'
import '@testing-library/jest-dom'
import dayjs from 'dayjs'

import { enableFetchMocks } from 'jest-fetch-mock'

test('RoomGraph', async () => {
  enableFetchMocks()
  fetchMock.mockResponseOnce(
    JSON.stringify([{ value: 0, updatedAt: dayjs(new Date(2020, 7, 31, 1, 0, 0, 0)) }]))
  const start = dayjs(new Date(2020, 7, 31, 0, 0, 0, 0))
  const end = dayjs(new Date(2020, 7, 31, 10, 0, 0, 0))

  const { getByRole } = render(<RoomGraph roomName={"test"} logType={"test"} start={start} end={end} />)
  await waitForElementToBeRemoved(getByRole("progressbar"))
  expect(screen.getByRole("listitem")).toBeInTheDocument()
  expect(fetchMock).toBeCalledWith(
    `/api/rooms/test/logs/test?start=${start.toISOString()}&end=${end.toISOString()}`)
})
