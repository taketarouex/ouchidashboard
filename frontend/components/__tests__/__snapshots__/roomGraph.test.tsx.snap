// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomGraph 1`] = `
<div>
  <div
    class="recharts-wrapper"
    style="position: relative; cursor: default; width: 400px; height: 400px;"
  >
    <svg
      class="recharts-surface"
      height="400"
      version="1.1"
      viewBox="0 0 400 400"
      width="400"
    >
      <defs>
        <clippath
          id="recharts1-clip"
        >
          <rect
            height="390"
            width="390"
            x="5"
            y="5"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</div>
`;
